library/std/src/net/parser.rs

```diff
@@ -299,34 +299,44 @@ impl FromStr for Ipv4Addr {
 impl FromStr for Ipv6Addr {
     type Err = AddrParseError;
     fn from_str(s: &str) -> Result<Ipv6Addr, AddrParseError> {
-        Parser::new(s).parse_with(|p| p.read_ipv6_addr())
+        Parser::new(s).parse_with(|p| p.read_ipv6_addr(), Kind::Ipv6Addr)
     }
 }
 
 #[stable(feature = "socket_addr_from_str", since = "1.5.0")]
 impl FromStr for SocketAddrV4 {
     type Err = AddrParseError;
     fn from_str(s: &str) -> Result<SocketAddrV4, AddrParseError> {
-        Parser::new(s).parse_with(|p| p.read_socket_addr_v4())
+        Parser::new(s).parse_with(|p| p.read_socket_addr_v4(), Kind::SocketAddrV4)
     }
 }
 
 #[stable(feature = "socket_addr_from_str", since = "1.5.0")]
 impl FromStr for SocketAddrV6 {
     type Err = AddrParseError;
     fn from_str(s: &str) -> Result<SocketAddrV6, AddrParseError> {
-        Parser::new(s).parse_with(|p| p.read_socket_addr_v6())
+        Parser::new(s).parse_with(|p| p.read_socket_addr_v6(), Kind::SocketAddrV6)
     }
 }
 
 #[stable(feature = "rust1", since = "1.0.0")]
 impl FromStr for SocketAddr {
     type Err = AddrParseError;
     fn from_str(s: &str) -> Result<SocketAddr, AddrParseError> {
-        Parser::new(s).parse_with(|p| p.read_socket_addr())
+        Parser::new(s).parse_with(|p| p.read_socket_addr(), Kind::SocketAddr)
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum Kind {
+    IpAddr,
+    Ipv4Addr,
+    Ipv6Addr,
+    SocketAddr,
+    SocketAddrV4,
+    SocketAddrV6,
+}
```

Comment:
I realize the type is private to this module, but I think `Kind` is too generic a name.

```suggestion
enum AddrKind {
    Ip,
    Ipv4,
    Ipv6,
    Socket,
    SocketV4,
    SocketV6,
}
```

Reply:
The amusing thing here is that I first wrote this as `AddrParseErrorKind`, then decided that was unnecessarily long and shortened it to `AddrKind` so that it matched what you have exactly, then decided I kinda liked the variants matching the type names, thereby fitting in with the whole matter of the diverse positioning of the `Addr` and capitalisation of the `v`. But yeah, looking back on it a day later, I think I prefer `AddrKind` again.